### PR TITLE
Move all rustc arguments to kani-driver

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Move all rustc arguments to kani-driver

1. It's good to have one source of truth.
2. This is needed for us to properly support caching compiler artifacts

### Resolved issues:

None. It's part of #2137 and needed for #2036.

### Related RFC:

Optional #ISSUE-NUMBER.

### Testing:

* How is this change tested? Current tests

* Is this a refactor change? Yes

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-driver/src/call_cargo.rs [hunk 1]
```diff
@@ -125,8 +113,8 @@ impl KaniSession {
                     .args(&target.to_args())
                     .args(&pkg_args)
                     .env("RUSTC", &self.kani_compiler)
-                    .env("RUSTFLAGS", "--kani-flags")
-                    .env("KANIFLAGS", &crate::util::join_osstring(&kani_args, " "))
+                    .env("KANIFLAGS", kani_args.join(" "))
+                    .env("CARGO_ENCODED_RUSTFLAGS", rustc_args.join(OsStr::new("\x1f")))
```
Comment (zhassan-aws) [inline kani-driver/src/call_cargo.rs:117]: Can you clarify what this does?

File: kani-driver/src/call_single_file.rs [hunk 2]
```diff
@@ -93,19 +85,61 @@ impl KaniSession {
...
         flags
     }
 
-    /// These arguments are arguments passed to kani-compiler that are `rustc` specific.
-    /// These are also used by call_cargo to pass as the env var KANIFLAGS.
+    /// This function generates all rustc configurations required by our goto-c codegen.
+    ///
+    /// We override the `std` library using the mechanism described here:
+    /// <https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/.E2.9C.94.20Globally.20override.20an.20std.20macro/near/268873354>
+    /// for more details.
```
Comment (zhassan-aws) [inline kani-driver/src/call_single_file.rs:105]: Remove this line.

File: scripts/std-lib-regression.sh [hunk 3]
```diff
@@ -50,8 +50,7 @@ cd std_lib_test
 
 # Add some content to the rust file including an std function that is non-generic.
 echo '
-#[kani::proof]
```
Comment (zhassan-aws) [inline scripts/std-lib-regression.sh:53]: Why was this removed?

File: scripts/std-lib-regression.sh [hunk 4]
```diff
@@ -68,7 +67,7 @@ echo "Starting cargo build with Kani"
 export RUST_BACKTRACE=1
 export RUSTC_LOG=error
 export KANIFLAGS="--goto-c --ignore-global-asm --reachability=legacy"
-export RUSTFLAGS="--kani-flags"
+export RUSTFLAGS="--kani-flags -C panic=abort"
```
Comment (zhassan-aws) [inline scripts/std-lib-regression.sh:70]: Why this change?

File: kani-driver/src/call_cargo.rs (hunk 1 above)
Comment (celinval) [reply kani-driver/src/call_cargo.rs:117]: It's the same as RUSTFLAGS. Per [their documentation](https://doc.rust-lang.org/cargo/reference/environment-variables.html), it is the recommended way. I don't know if it matters much here, but I figured why not:

> This string is split by whitespace; for a more robust encoding of multiple arguments, see CARGO_ENCODED_RUSTFLAGS.

File: scripts/std-lib-regression.sh (hunk 3 above)
Comment (celinval) [reply scripts/std-lib-regression.sh:53]: So we don't have to manually inject Kani dependency here. Since our goal for this script is to make sure we can generate code for the standard library, this seems much easier to maintain than duplicating a bunch of the compiler flags here.

File: scripts/std-lib-regression.sh (hunk 4 above)
Comment (celinval) [reply scripts/std-lib-regression.sh:70]: I had to change this because in this script we are invoking `kani-compiler` via `cargo build` not via `cargo kani`. This is the only relevant flag for the codegen to go through.

File: kani-driver/src/call_cargo.rs (hunk 1 above)
Comment (zhassan-aws) [reply kani-driver/src/call_cargo.rs:117]: Please add a link to the documentation

**Review by zhassan-aws — APPROVED** [review]

File: kani-driver/src/call_single_file.rs (hunk 2 above)
Comment (zhassan-aws) [reply kani-driver/src/call_single_file.rs]: Sorry, I just meant the last line of the comment: 
```
/// for more details.
```
to fix the sentence and not the entire comment. The rest of the comment is still relevant.
